content/blog/2022-05-23-vectorization/index.md

```diff
@@ -0,0 +1,238 @@
+
+```bril
+w: int = add a b;
+x: int = add b c;
+y: int = add a c;
+z: int = add d d;
+```
+
+would be a valid run of vectorizable instructions, while the similar sequence:
+
+```bril
+w: int = add a b;
+x: int = add b w;
+y: int = add a c;
+z: int = add d d;
+```
+
+would not be a valid run of vectorizable instructions because `w` is defined and then used in the same basic block.
+
+Finally, the third stage is to run the vectorization algorithms on the runs of vectorizable instructions, and create vectors, which then replace code in the basic block. Vectors are only created when each run has two or more instructions in it.
+
+### Preprocessing
+
+Preprocessing is applied to the Bril code to allow for further optimization potential. Assuming no memory operations are present, the first preprocessing optimizations applied are simple dead code elimination (DCE) and loop invariant code motion (LICM). Removing obviously dead instructions will help simplify the code. As a side note, the condition for non-present memory conditions is because my implementations do not handle memory operations for DCE or LICM. Also, local value numbering (LVN) would have been used, but my version of LVN had a bug, causing me to exclude using LVN.
+
+I then inline as much code as possible, to create larger basic blocks. To do so, I created a call graph of all calls, then topologically sort the call graph, and copy the body of callee functions into callers, stitching together code as appropriate.
+
+Next, loops in the code are fully unrolled, if possible. Unrolling only applies to loops that have one iteration variable that is increasing by 1 or decreasing by 1. Further, the loop header must have exactly one branch condition, depending on the iteration variable. The loop also must have only one exit, from the branch condition in the header. By unrolling the code, loops get expanded into long sequences of straight-line code without branches or jumps. Having straight line code allows for long sequences of instructions that can be identified as vectorizable, and the possibility that a vector may be used several times.
+
+After unrolling, stores are then moved as late in the basic block as possible. Store movement depends on alias analysis, because a store can only move past another store or another load, assuming no aliasing. Moving stores later in the basic block allows for a greater opportunity of finding vectors, because vectors have to end when a store is identified. In the same vein, `const` instructions and `id` instructions are moved as early as possible, while `print` instructions are moved to be as far down in a basic block as possible.
```

Comment:
Since it will inevitably come up as a question, maybe you want to head off the reader asking "but what alias analysis did you implement?" by giving a quick preview or saying that you'll address this later?

Reply:
I've added a paragraph describing the general mechanism for alias analysis.